feat: dev command for managing extensions

n0Oo0Oo0b opened the pull request:
This allows for hot-reloading cogs in extensions (i.e. changes can be reloaded without having to restart the whole bot).

Usage:

- `<prefix>ext list`: shows a list of loaded extensions
- `<prefix>ext load <ext_name>`: loads a new extension (not loaded yet)
- `<prefix>ext unload <ext_name>`: unloads a loaded extension
- `<prefix>ext reload <ext_name>`: reloads an already loaded extension

Extensions are assumed to be in `cogs/`. `<ext_name>` can be just the name of the extension (e.g. `dev`) or the full path (e.g. `cogs.dev`).

Note that changes in app commands may have to be synced separately (`sync *`) after (re-)loading an extension.

n0Oo0Oo0b committed:
feat: dev command for managing extensions

n0Oo0Oo0b committed:
fix: limit ext command to administrators

We probably want a better check for this, but I'm just adding this now in case it reaches production

n0Oo0Oo0b committed:
style: remove trailing whitespace

ilikecubesnstuff commented on cogs/dev.py:
```diff
@@ -83,6 +89,37 @@ async def sync(self, ctx: commands.Context, guilds: commands.Greedy[discord.Obje
+    async def ext_load(self, ctx, ext_name: ext_converter):
+        """Loads an extension"""
+        await self.bot.load_extension(ext_name)
+        ctx.send("Success!")
```
Await?

ilikecubesnstuff commented on cogs/dev.py:
```diff
@@ -83,6 +89,37 @@ async def sync(self, ctx: commands.Context, guilds: commands.Greedy[discord.Obje
         await ctx.send(f"Synced the tree to {ret}/{len(guilds)}.")
 
+    @commands.group(aliases=['ext'])
+    @commands.has_permissions(administrator=True)
```
Honestly, I thought this was already a cog check... Could you change this into a cog check? Also check against `manage_guild` instead of `administrator`, since I think that is the highest permission that the `SysAdmin` role has in the server.

ilikecubesnstuff commented on cogs/dev.py:
```diff
@@ -83,6 +89,37 @@ async def sync(self, ctx: commands.Context, guilds: commands.Greedy[discord.Obje
+    async def ext_reload(self, ctx, ext_name: ext_converter):
+        """Reloads an extension"""
+        await self.bot.reload_extension(ext_name)
+        await ctx.send("Success!")
```
Make the responses more descriptive—say "Successfully loaded/unloaded/reloaded `<ext name>`."

ilikecubesnstuff reviewed (commented):
Thank you for implementing these! I have left a few comments, but I also wanted to ask: are these commands ever to be used in production? Or is this just for development convenience?

n0Oo0Oo0b committed:
fix: add cog check to `cogs.dev`

Required permission is `manage_guild` as per https://github.com/ib-ai/ib.py/pull/34#discussion_r1606790860 (highest perm that SysAdmins have)

n0Oo0Oo0b committed:
fix: more descriptive response for `extension load/unload/reload`

n0Oo0Oo0b commented:
> are these commands ever to be used in production? Or is this just for development convenience?

I'm adding this mainly for development convenience but I could see this being used in production too (e.g. if a vulnerability is found in a specific cog then the rest of the bot can run as normal while it gets patched)

n0Oo0Oo0b committed:
fix: use check from `utils.checks`

NathanealV commented on cogs/dev.py:
```diff
@@ -6,18 +6,28 @@
+from utils.checks import cogify, admin_command
+
+
+def ext_converter(argument: str):
+    if not argument.startswith("cogs."):
+        argument = f"cogs.{argument}"
+    return argument
```
Could you `.trim()` the argument before processing?

NathanealV commented on cogs/dev.py:
```diff
@@ -83,6 +93,36 @@ async def sync(self, ctx: commands.Context, guilds: commands.Greedy[discord.Obje
+    @commands.group(aliases=['ext'])
+    async def extensions(self, ctx: commands.Context):
+        """Utilities for extensions"""
+        if ctx.invoked_subcommand is None:
+            await ctx.send_help("ext")
```
Didn't know `ctx.send_help` existed - we've been using `await available_subcommands(ctx)` for command groups. To ensure consistency for now, would prefer to use that function here.

Maybe re-working that function to include `ctx.send_help` can be done in the future (in #36)

NathanealV commented on cogs/dev.py:
```diff
@@ -83,6 +93,36 @@ async def sync(self, ctx: commands.Context, guilds: commands.Greedy[discord.Obje
+        if ctx.invoked_subcommand is None:
+            await ctx.send_help("ext")
+
+    @extensions.command(name="list")
+    async def ext_list(self, ctx):
+        """Lists currently loaded extensions"""
+        ext_list = "\n".join(f"- {ext}" for ext in self.bot.extensions)
+        await ctx.send(f"List of loaded extensions:\n{ext_list}")
+
+    @extensions.command(name="load")
+    async def ext_load(self, ctx, ext_name: ext_converter):
+        """Loads an extension"""
+        await self.bot.load_extension(ext_name)
+        await ctx.send(f"Successfully loaded extension `{ext_name}`")
+
+    @extensions.command(name="unload")
+    async def ext_unload(self, ctx, ext_name: ext_converter):
+        """Unloads an extensions"""
+        await self.bot.unload_extension(ext_name)
+        await ctx.send(f"Successfully unloaded extension `{ext_name}`")
+
+    @extensions.command(name="reload")
+    async def ext_reload(self, ctx, ext_name: ext_converter):
+        """Reloads an extension"""
+        await self.bot.reload_extension(ext_name)
+        await ctx.send(f"Successfully reloaded extension `{ext_name}`")
```
Just a couple general comments here:
- Include the type for `ctx` within function parameters (`commands.Context`)
- Format the doc-strings to be like the other commands for consistency (newlines, punctuation) + typo in `"""Unloads an extensions"""`
- Add `logger.debug` statements when loading/unloading/reloading extensions (see `cogs/filter.py` for example)

n0Oo0Oo0b replied:
I decided to go for `logging.info` instead of `logging.debug` since I think it makes more sense.

NathanealV reviewed (changes requested):
Thank you for the PR! Should only be a couple quick changes - please see my PR comments.

n0Oo0Oo0b committed:
fix: apply changes from code review

Changes made:
- Strip extension names before processing
- Use existing `utils.commands.available_subcommands` instead of `ctx.send_help` for consistency (this may be changed back later in #36)
- Typehints for `ctx`
- Docstring formatting consistency
- Logging of load/unload/reloading of extensions

NathanealV reviewed (dismissed):
LGTM!

n0Oo0Oo0b committed:
docs: update docstring for sync command

ilikecubesnstuff reviewed (dismissed):
LGTM!

n0Oo0Oo0b committed:
Merge branch 'refs/heads/rewrite' into extensions-command

# Conflicts:
#	cogs/dev.py
